### ISSUE TITLE
Any tips for package provider not working

### ISSUE THREAD

**Issue by sgryphon** [ethanbergstrom/Chocolatier#8]
Trying the first few commands in the readme just generates errors:

```
PS C:\Windows\System32> Find-PackageProvider Chocolatier -verbose
VERBOSE: Acquiring providers for assembly: C:\program files\powershell\7\Modules\PackageManagement\coreclr\netstandard2.0\Microsoft.PackageManagement.ArchiverProviders.dll
VERBOSE: Acquiring providers for assembly: C:\program files\powershell\7\Modules\PackageManagement\coreclr\netstandard2.0\Microsoft.PackageManagement.CoreProviders.dll
VERBOSE: Acquiring providers for assembly: C:\program files\powershell\7\Modules\PackageManagement\coreclr\netstandard2.0\Microsoft.PackageManagement.MetaProvider.PowerShell.dll
VERBOSE: Acquiring providers for assembly: C:\program files\powershell\7\Modules\PackageManagement\coreclr\netstandard2.0\Microsoft.PackageManagement.NuGetProvider.dll
VERBOSE: Using the provider 'Bootstrap' for searching packages.
VERBOSE: Finding the package 'Bootstrap::FindPackage' 'Chocolatier','','','''.
VERBOSE: Using the provider 'PowerShellGet' for searching packages.
VERBOSE: The -Repository parameter was not specified.  PowerShellGet will use all of the registered repositories.
VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.
VERBOSE: The specified Location is 'https://www.powershellgallery.com/api/v2' and PackageManagementProvider is 'NuGet'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Chocolatier'' for ''.
VERBOSE: Total package yield:'1' for the specified package 'Chocolatier'.
Find-PackageProvider: No match was found for the specified search criteria and package name 'Chocolatier'. Try Get-PackageSource to see all available registered package sources.
```

And trying to install:

```
PS C:\Windows\System32> Install-PackageProvider Chocolatier -verbose
VERBOSE: Using the provider 'Bootstrap' for searching packages.
VERBOSE: Finding the package 'Bootstrap::FindPackage' 'Chocolatier','','','''.
VERBOSE: Using the provider 'PowerShellGet' for searching packages.
VERBOSE: The -Repository parameter was not specified.  PowerShellGet will use all of the registered repositories.
VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.
VERBOSE: The specified Location is 'https://www.powershellgallery.com/api/v2' and PackageManagementProvider is 'NuGet'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='Chocolatier'' for ''.
VERBOSE: Total package yield:'1' for the specified package 'Chocolatier'.
Install-PackageProvider: No match was found for the specified search criteria for the provider 'Chocolatier'. The package provider requires 'PackageManagement' and 'Provider' tags. Please check if the specified package has the tags.
```

Several links suggested an issue with TLS, or with not trusting the package sources, so I set TLS1.2 and trusted both sources:

```
PS C:\Windows\System32> [Net.ServicePointManager]::SecurityProtocol
Tls12
PS C:\Windows\System32> Get-PackageSource

Name                             ProviderName     IsTrusted  Location
----                             ------------     ---------  --------
nuget.org                        NuGet            True       https://api.nuget.org/v3/index.json
PSGallery                        PowerShellGet    True       https://www.powershellgallery.com/api/v2
```

This is using the latest version of PowerShell, on a Windows 10 machine:

```
PS C:\Windows\System32> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

I did download the package manually, and checked it has the needed tags.

Also note that the verbose seems to indicate the package is found, just throws an error at the end (and if I pass `-AllVersions` it has the line `VERBOSE: Total package yield:'4' for the specified package 'Chocolatier'.` but still then gives the error.

And, `Find-Package Chocolatier` does actually find the package (just Find-PackageProvider does not).

**Comment by sgryphon** [comment]
More information. The package can be found as a normal package, so I tried installing it this way:

```
Find-Package Chocolatier -Verbose
Install-Module Chocolatier -Verbose
Import-PackageProvider Chocolatier
Get-PackageProvider
``` 

That worked, i.e. it is now listed in Get-PackageProvider.

Trying an actual package command, such as list sources, gave the option to install choco:

```
PS C:\Windows\System32> Get-PackageSource

Name                             ProviderName     IsTrusted  Location
----                             ------------     ---------  --------
nuget.org                        NuGet            True       https://api.nuget.org/v3/index.json
PSGallery                        PowerShellGet    True       https://www.powershellgallery.com/api/v2

Choco.exe is required to continue
Chocolatier is built on Choco.exe. Do you want Chocolatier to install Choco.exe from
'https://chocolatey.org/install.ps1' now?
[Y] Yes  [N] No  [S] Suspend  [?] Help (default is "Y"): y
```

No package source, but I thought I would try finding something anyway:

```
Find-Package -Name  nodejs -ProviderName Chocolatier -Verbose

PS C:\Windows\System32> Find-Package -ProviderName Chocolatier -name  nodejs
WARNING: Cannot bind argument to parameter 'SourceName' because it is an empty string.
WARNING: You cannot call a method on a null-valued expression.
WARNING: Cannot find an overload for "ErrorRecord" and the argument count: "4".
WARNING: The scope number '1' exceeds the number of active scopes.
Parameter name: scopeID
Actual value was 1.
WARNING: Cannot validate argument on parameter 'ExceptionMessage'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
WARNING: The expression after '&' in a pipeline element produced an object that was not valid. It must result in a command name, a script block, or a CommandInfo object.
WARNING: Cannot find a variable with the name 'AdditionalArgs'.
WARNING: File C:\Users\Sly Gryphon\AppData\Local\Temp\chocolatey\chocInstall\tools\chocolateyInstall.ps1 cannot be loaded because running scripts is disabled on this system. For more information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
WARNING: Cannot process argument transformation on parameter 'ErrorCategory'. Cannot convert value "InvalidOperationException" to type "System.Management.Automation.ErrorCategory". Error: "Unable to match the identifier name InvalidOperationException to a valid enumerator name. Specify one of the following enumerator names and try again:
NotSpecified, OpenError, CloseError, DeviceError, DeadlockDetected, InvalidArgument, InvalidData, InvalidOperation, InvalidResult, InvalidType, MetadataError, NotImplemented, NotInstalled, ObjectNotFound, OperationStopped, OperationTimeout, SyntaxError, ParserError, PermissionDenied, ResourceBusy, ResourceExists, ResourceUnavailable, ReadError, WriteError, FromStdErr, SecurityError, ProtocolError, ConnectionError, AuthenticationError, LimitsExceeded, QuotaExceeded, NotEnabled"
Find-Package: No match was found for the specified search criteria and package name 'nodejs'. Try Get-PackageSource to see all available registered package sources.
```

Trying to register the default Chocolately source also got an error the first time (and nothing listed in source):

```
PS C:\Windows\System32> Register-PackageSource -Name Chocolatey -Location https://chocolatey.org/api/v2/ -Trusted -ProviderName Chocolatier -Verbose
Register-PackageSource: Cannot validate argument on parameter 'ExceptionMessage'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
```

**Comment by sgryphon** [comment]
Looking at the first error, with the failed install, I changed the execution policy and then ran the script directly (the path would depend on the current user).

```
Set-ExecutionPolicy Unrestricted         
& "C:\Users\Sly Gryphon\AppData\Local\Temp\chocolatey\chocInstall\tools\chocolateyInstall.ps1"
```

Closing the admin powershell console, and opening a new one, still not working:

```
PS C:\Windows\System32> Find-Package -Name  nodejs -ProviderName Chocolatier -Verbose
VERBOSE: Using the provider 'Chocolatier' for searching packages.
WARNING: Cannot find type [Microsoft.PackageManagement.MetaProvider.PowerShell.SoftwareIdentity]: verify that the assembly containing this type is loaded.
WARNING: Cannot find a variable with the name 'AdditionalArgs'.
Find-Package: No match was found for the specified search criteria and package name 'nodejs'. Try Get-PackageSource to see all available registered package sources.
```

And registering a source generates a similar error:

```
PS C:\Windows\System32> Register-PackageSource -Name Chocolatey -Location https://chocolatey.org/api/v2/ -Trusted -ProviderName Chocolatier -Verbose
Register-PackageSource: Cannot find type [Microsoft.PackageManagement.MetaProvider.PowerShell.PackageSource]: verify that the assembly containing this type is loaded.
```

**Comment by ethanbergstrom** [comment]
Hey @sgryphon

It looks like you might be trying this with PowerShell 7? I'll be honest, I haven't tried the module above 5.1. 

It looks like PackageManagement under PS7 implements Find-/Install-PackageProvider differently than on 5.1. When I run just straight `Find-PackageProvider` under PowerShell 7, I only get three results from PSGallery, while under PowerShell 5.1 I get 22 results. 

Likewise, when I run commands like `Get-PackageSource` or `Find-Package` under PowerShell 7, I get the same results you do, whereas under PowerShell 5.1 everything works as expected. 

Ostensibly, I'm running PackageManagement 1.4.6 in both PowerShell 7 and 5.1, but they seem to be behaving differently. Interestingly, if I start with a fresh PowerShell 7 instance and import PackageManagement using the `-UseWindowsPowerShell` argument, it seems to import the module through a compatibility mode. Once PackageManagement is imported that way, Chocolatier seems to work as expected. 

Based on this, there seems to be some pretty significant breaking changes in either PowerShell 7 or the flavor of PackageManagement shipped for PowerShell 7 that renders my module no longer compatible, which I'd have to spend time researching and reworking. 

In the interim, I would say either use Chocolatier with PowerShell 5.1, or first import the PackageManagement module in PowerShell 7 with the compatibility flag:
`Import-Module packagemanagement -UseWindowsPowerShell`

**Comment by ethanbergstrom** [comment]
Following up on this. After playing around with PowerShell 7 a bit, the issues I ran into seemed to be due to a bug in the initial GA release of PS7 in which it was unable to properly import certain DLLs (namely those shipped in Microsoft's own PackageManagement module).

If it is indeed the culprit bug, it was just fixed a few days ago in the latest 7.1 preview: https://github.com/PowerShell/PowerShell/pull/12203

In my tests, PS 7.1 preview 2 loads the module just fine without any special compatibility flags, however it still doesn't find the provider in PSGallery, which I think may be due to a nonpassive change in search behavior with module tags. Will try to push a new version that has the tagging required for it to show up in `Find-PackageProvider` / `Install-PackageProvider`

PowerShell 7:
```powershell
PS C:\Users\ethan> find-package firefox -providername chocolatier
WARNING: Cannot find type [Microsoft.PackageManagement.MetaProvider.PowerShell.SoftwareIdentity]: verify that the assembly containing this type is loaded.
WARNING: Cannot find a variable with the name 'AdditionalArgs'.
Find-Package: No match was found for the specified search criteria and package name 'firefox'. Try Get-PackageSource to see all available registered package sources.
```

PowerShell 7.1 Preview:
```powershell
PS C:\Users\ethan> find-package firefox -providername chocolatier

Name                           Version          Source           Summary
----                           -------          ------           -------
Firefox                        75.0             chocolatey
```

**Comment by ethanbergstrom** [comment]
Hey @sgryphon,

I think this should be fixed now with [v1.0.5](https://github.com/ethanbergstrom/Chocolatier/compare/v1.0.4...v1.0.5#diff-4ac32a78649ca5bdd8e0ba38b7006a1e). 

The provider can now be found by both PowerShell 5.1 and 7.1, and loads correctly in both:

```powershell
PS C:\Users\ethan> $PSVersionTable.PSVersion.ToString()
7.1.0-preview.2

PS C:\Users\ethan> Find-PackageProvider
Name                           Version          Source           Summary
----                           -------          ------           -------
PowerShellGet                  2.2.4.1          PSGallery        PowerShell module with commands f…
ContainerImage                 0.6.4.0          PSGallery        This is a PackageManagement provi…
NanoServerPackage              1.0.1.0          PSGallery        A PackageManagement provider to  …
Chocolatier                    1.0.5            PSGallery        Package Management (OneGet) provi…

PS C:\Users\ethan> install-packageprovider chocolatier -Force
WARNING: The version '1.4.7' of module 'PackageManagement' is currently in use. Retry the operation after closing the applications.

Name                           Version          Source           Summary
----                           -------          ------           -------
Chocolatier                    1.0.5            PSGallery        Package Management (OneGet) provi…

PS C:\Users\ethan> install-package cpu-z -providername chocolatier -acceptlicense | uninstall-package

Name                           Version          Source           Summary
----                           -------          ------           -------
cpu-z                          v1.92            chocolatey
```

Please note that I had to use PowerShell 7.1 Preview 2 in order for it to work. The provider is not going to work with PowerShell 7 due to a defect that breaks how PackageManagement loads it's own DLLs. I would imagine quite a few package providers are broken on PS7 as a result. I've updated the module manifest so that it will refuse to load into a CoreCLR version of PowerShell less than 7.1:

```powershell
PS C:\Users\ethan> $PSVersionTable.PSVersion.ToString()
7.0.0

PS C:\Users\ethan> find-packageprovider
Name                           Version          Source           Summary
----                           -------          ------           -------
PowerShellGet                  2.2.4.1          PSGallery        PowerShell module with commands f…
ContainerImage                 0.6.4.0          PSGallery        This is a PackageManagement provi…
NanoServerPackage              1.0.1.0          PSGallery        A PackageManagement provider to  …
Chocolatier                    1.0.5            PSGallery        Package Management (OneGet) provi…

PS C:\Users\ethan> install-packageprovider chocolatier -force
WARNING: The version '1.4.7' of module 'PackageManagement' is currently in use. Retry the operation after closing the applications.
Install-PackageProvider: Index (zero based) must be greater than or equal to zero and less than the size of the argument list.

PS C:\Users\ethan> get-packageprovider -ListAvailable

Name                     Version          DynamicOptions
----                     -------          --------------
NuGet                    3.0.0.1          Destination, ExcludeVersion, Scope, SkipDependencies, He…
PowerShellGet            2.2.3.0          PackageManagementProvider, Type, Scope, AllowClobber, Sk…

PS C:\Users\ethan> import-module chocolatier
Import-Module: The version of PowerShell on this computer is '7.0.0'. The module 'C:\Users\ethan\OneDrive\Documents\PowerShell\Modules\chocolatier\1.0.5\chocolatier.psd1' requires a minimum PowerShell version of '7.1.0' to run. Verify that you have the minimum required version of PowerShell installed, and then try again.
```

When you get the chance, please try this out with the [latest PowerShell 7.1 preview](https://github.com/PowerShell/PowerShell/releases) and see if it works for you. 

Thanks!

**Comment by ethanbergstrom** [comment]
Just an update on this issue - it looks like [last night's release of PowerShell 7.0.1](https://github.com/PowerShell/PowerShell/releases/tag/v7.0.1) contains the same correction as in the 7.1 preview that inhibited PackageManagement from working properly. 

I just pushed [v1.0.6](https://github.com/ethanbergstrom/Chocolatier/compare/v1.0.5...v1.0.6) to the provider to extend it's compatibility range to PowerShell 7.0.1. It seems to be working well in my tests. 

When you get the change, please update to PowerShell 7.0.1 and give the provider another shot.
